### DISCUSSION TITLE
classes/class_subviewport

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#214] [User-contributed notes]
# classes/class_subviewport

Inherits: Viewport< Node< Object An interface to a game world that doesn't create a window or draw to the screen directly. Description: SubViewport Isolates a rectangular region of a scene to be di...

https://docs.godotengine.org/en/stable/classes/class_subviewport.html

**Comment by jupiterbjy** [comment]
Don't forget to turn on Physics>Object Picking(`.physics_object_picking`) when using `mouse_entered/exited` signals inside `SubViewport` - cause it's off by default.

**Reply by jian-kim** [reply]
It took me 2 hours of digging and 80% through creating a bug report to find this, so thank you.

I feel like this should be documented somewhere easier to find

**Reply by jupiterbjy** [reply]
yeah it's real hard to search. I thought of writting issue for more clear documentation but no good wording came up so I just wrote it here haha

**Reply by elamaunt** [reply]
Clicks will also fail if the Mouse Filter is not set to Ignore on other controls that overlap the SubViewportContainer

**Comment by Emaxoso** [comment]
If you are using multiple SubViewports in your scene, remember to enable the 'own_world_3d' flag; otherwise, all the viewports will render the same content!

**Reply by thygrrr** [reply]
This is usually what you want, multiple views into the same world. Use camera culling masks to filter out specific nodes.

**Comment by joethephish** [comment]
I find the following terminology slightly easier to understand for the three size params:

- `size`: the literal pixel size of the render texture
- `size_2d_override`: the size of the 2D layout space for canvas items rendered by this viewport. 
- `size_2d_override_stretch`: when `size_2d_override` differs from `size`, this controls whether the 2D content should be stretched to fill the full texture, or rendered at the override size within it (which may lead to borders or clipping).

For example, with `size_2d_override_stretch` enabled, setting `size_2d_override` to 1080p and `size` to 4K means the UI will be laid out as if at 1080p, but rendered at 4K — effectively giving you supersampled 2D visuals.

That's my understanding anyway, very happy to be corrected! Would be great if the official docs could be clarified to reflect this.

**Reply by Calinou** [reply]
> For example, with size_2d_override_stretch enabled, setting size_2d_override to 1080p and size to 4K means the UI will be laid out as if at 1080p, but rendered at 4K — effectively giving you supersampled 2D visuals.

Yes, this is correct :slightly_smiling_face: 

However, in this case, remember that 2D visuals will be pixelated/blurry unless drawing is vector-based, or sprites are oversampled somehow (e.g. using font oversampling or SVGTexture in 4.5).

**Reply by joethephish** [reply]
Right yeah. I guess in my case a lot of stuff is automatically vector based, like if I'm using a lot of panels etc, or if I was already using oversized textures that were shrunk down for this very reason.

Does Godot handle text rendering correctly for this case, using SDF or similar?

**Reply by Calinou** [reply]
> Does Godot handle text rendering correctly for this case, using SDF or similar?

Font oversampling is enabled by default, but you can also opt into [MSDF font rendering](https://docs.godotengine.org/en/stable/tutorials/ui/gui_using_fonts.html#doc-using-fonts-msdf) which doesn't need font oversampling to remain crisp.

**Comment by dukeman74** [comment]
I was trying to fix an issue that trying to draw/move objects to be in line with objects that are in a subviewport works easily, using get_viewport_transform(), however, if you move the camera in the suviewport's world, they lag behind by a frame.

After some unfruitful research, and much debugging and trial and error, the problem is that the viewport transform is not updated when the camera's transform is updated.
to work around this, I have the camera track its own real time viewport_transform variable
this one matches the engine one in most cases, but sometimes the floats differ by a miniscule amount.  I would like to know why, but it doesn't matter for practical use.

set this after the camera is moved, and make sure that the script running to position/draw outside of this viewport's world has a priority/tree position that causes it to run after this camera's script.
```GDScript
real_time_transform = Transform2D(0,zoom,0,Vector2(viewport.size)/2 - position*zoom )
```
my camera does not rotate or skew, but those numbers would be simple to drop in.

to get the projected position, just multiply the object in the viewport's world 's global_position by this transform

**Comment by thygrrr** [comment]
# PURPLE VIEWPORT TEXTURE?
**Godot 4.3+, all the way through 4.6.0-rc1**
If your ViewPortTexture looks fashion doll purple or checkered purple in game, *even though it works in editor*, check your MeshInstance or similar that is rendering it, and the materials. 
- Mesh Resource must be set `Local to Scene`
- Material Resource must be set `Local to Scene`
- ViewportTexture Resource must be set `Local to Scene`
- MeshInstance Skeleton must be set (it gets cleared after reparenting nodes)

After making these changes, you must **close and re-open the scene** for them to work.
